mbzirc_ign/src/WaveVisual.cc

```diff
@@ -0,0 +1,351 @@
+#include "ignition/gazebo/rendering/RenderUtil.hh"
+#include "ignition/gazebo/Util.hh"
+
+using namespace ignition;
+using namespace gazebo;
+using namespace systems;
+
+class ignition::gazebo::systems::WaveVisualPrivate
+{
+  /// \brief Path to vertex shader
+  public: std::string vertexShaderUri;
+
+  /// \brief Path to fragment shader
+  public: std::string fragmentShaderUri;
+
+  /// \brief Mutex to protect sim time updates.
+  public: std::mutex mutex;
+
+  /// \brief Connection to pre-render event callback
+  public: ignition::common::ConnectionPtr connection{nullptr};
+
+  /// \brief Name of visual this plugin is attached to
+  public: std::string visualName;
+
+  /// \brief Pointer to visual
+  public: rendering::VisualPtr visual;
+
+  /// \brief Material used by this visual
+  public: rendering::MaterialPtr material;
+
+  /// \brief Pointer to scene
+  public: rendering::ScenePtr scene;
+
+  /// \brief Entity id of the visual
+  public: Entity entity = kNullEntity;
+
+  /// \brief Current sim time
+  public: std::chrono::steady_clock::duration currentSimTime;
+
+  /// \brief Path to model
+  public: std::string modelPath;
+
+  /// \brief Wavefield for computing wave params
+  public: Wavefield wavefield;
+
+  /// \brief Indicate whether the shader params have been set or not
+  public: bool paramsSet = false;
+
+  /// \brief All rendering operations must happen within this call
+  public: void OnUpdate();
+};
+
+/////////////////////////////////////////////////
+WaveVisual::WaveVisual()
+    : System(), dataPtr(std::make_unique<WaveVisualPrivate>())
+{
+}
+
+/////////////////////////////////////////////////
+WaveVisual::~WaveVisual()
+{
+}
+
+/////////////////////////////////////////////////
+void WaveVisual::Configure(const Entity &_entity,
+               const std::shared_ptr<const sdf::Element> &_sdf,
+               EntityComponentManager &_ecm,
+               EventManager &_eventMgr)
+{
+  IGN_PROFILE("WaveVisual::Configure");
+  // Ugly, but needed because the sdf::Element::GetElement is not a const
+  // function and _sdf is a const shared pointer to a const sdf::Element.
+  auto sdf = const_cast<sdf::Element *>(_sdf.get());
+
+  if (!sdf->HasElement("wavefield"))
+  {
+    ignerr << "<wavefield> parameter is missing " << std::endl;
+    return;
+  }
+  if (!sdf->HasElement("shader"))
+  {
+    ignerr << "<shader> parameter is missing " << std::endl;
+    return;
+  }
+
+  this->dataPtr->wavefield.Load(_sdf);
+
+  if (this->dataPtr->modelPath.empty())
+  {
+    auto modelEntity = topLevelModel(_entity, _ecm);
+    this->dataPtr->modelPath =
+        _ecm.ComponentData<components::SourceFilePath>(modelEntity).value();
+  }
+
+  // parse path to shaders
+  sdf::ElementPtr shaderElem = sdf->GetElement("shader");
+  if (!shaderElem->HasElement("vertex") ||
+      !shaderElem->HasElement("fragment"))
+  {
+    ignerr << "<shader> must have <vertex> and <fragment> sdf elements"
+           << std::endl;
```

Comment:
Do you need to return here?

Reply:
yep done. 6b1390777afb53df8a9011cefd6b4649eb19baf9